qutip/lattice.py

```diff
@@ -0,0 +1,1301 @@
+#       from this software without specific prior written permission.
+#
+#    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+#    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###############################################################################
+
+
+__all__ = ['Lattice1d', 'cell_structures']
+
+from scipy.sparse import (csr_matrix)
+from qutip import (Qobj, tensor, basis, qeye, isherm, sigmax, sigmay, sigmaz)
+
+import numpy as np
+
+try:
+    import matplotlib.pyplot as plt
+except:
+    pass
+
+
+def Hamiltonian_2d(base_h, inter_hop_x, inter_hop_y, nx_units=1, PBCx=0,
+                   ny_units=1, PBCy=0, cell_num_site=1, length_for_site=1):
+    """
+    Returns the Hamiltonian as a csr_matrix from the specified parameters for a
+    2d space of sites in the x-major format.
+
+    Parameters
+    ==========
+    base_h : qutip.Qobj
+        The Hamiltonian matrix of the unit cell
+    inter_hop_x : list of Qobj/ qutip.Qobj
+        The matrix coupling between cell i to the one at cell i+\hat{x}
+    inter_hop_y : list of Qobj/ qutip.Qobj
+        The matrix coupling between cell i to the one at cell i+\hat{y}
+    site_dof_config : list
+        The list of two numbers, namely the number of sites and the nuber of
+        degrees of freedom per site of the unit cell
+    nx_units : int
+        The length of the crystal in the x direction in units of unit cell
+        length
+    PBCx : int
+        The indicator of periodic(1)/hardwall(0) boundary condition along
+        direction x
+    ny_units : int
+        The length of the crystal in the x direction in units of unit cell
+        length
+    PBCy : int
+        The indicator of periodic(1)/hardwall(0) boundary condition along
+        direction y
+
+    Returns
+    -------
+    Hamt : csr_matrix
+        The 2d Hamiltonian matrix for the specified parameters.
+    """
+    xi_len = len(inter_hop_x)
+    yi_len = len(inter_hop_y)
+    (x0, y0) = np.shape(inter_hop_x[0])
+
+    if inter_hop_y == []:
+        x1 = 0
+        y1 = 0
+    else:
+        (x1, y1) = np.shape(inter_hop_y[0])
+    (xx, yy) = np.shape(base_h)
+
+    row_ind = np.array([])
+    col_ind = np.array([])
+    data = np.array([])
+
+    NS = cell_num_site*length_for_site
```

Comment:
`NS`: I would avoid using capital letters for variables that are not classes. Also, aim at self-explaining names for variables. `NS` could be `site_number` or what most appropriate.